File: SundeeFundee/Packages/SundeeFundeeShared/Sources/SundeeFundeeShared/Models/ExerciseCatalog.swift
import Foundation

public enum WeightliftingExerciseCatalog {

    public enum Category: String, CaseIterable, Sendable {
        case squat    = "Squat"
        case hinge    = "Hip Hinge"
        case press    = "Press"
        case pull     = "Pull"
        case carry    = "Carry"
        case olympic  = "Olympic Weightlifting"
    }

    public struct Entry: Identifiable, Sendable {
        public let id: String
        public let category: Category
        public init(id: String, category: Category) {
            self.id = id; self.category = category
        }
    }

    public static let all: [Entry] = [
        // Squat
        Entry(id: "Back Squat",            category: .squat),
        Entry(id: "Front Squat",           category: .squat),
        Entry(id: "Safety Bar Squat",      category: .squat),
        Entry(id: "Box Squat",             category: .squat),
        Entry(id: "Pause Squat",           category: .squat),
        Entry(id: "Goblet Squat",          category: .squat),
        // Hip hinge
        Entry(id: "Conventional Deadlift (No Straps)", category: .hinge),
        Entry(id: "Conventional Deadlift (With Straps)", category: .hinge),
        Entry(id: "Romanian Deadlift (No Straps)",     category: .hinge),
        Entry(id: "Romanian Deadlift (With Straps)",     category: .hinge),
        Entry(id: "Sumo Deadlift (No Straps)",         category: .hinge),
        Entry(id: "Sumo Deadlift (With Straps)",         category: .hinge),
        Entry(id: "Trap Bar Deadlift (No Straps)",     category: .hinge),
        Entry(id: "Trap Bar Deadlift (With Straps)",     category: .hinge),
        Entry(id: "Good Morning",          category: .hinge),
        Entry(id: "Hip Thrust",            category: .hinge),
        // Press
        Entry(id: "Flat Barbell Bench Press",  category: .press),
        Entry(id: "Incline Barbell Bench Press", category: .press),
        Entry(id: "Strict Press",          category: .press),
        Entry(id: "Push Press",            category: .press),
        Entry(id: "Dumbbell Bench Press",  category: .press),
        Entry(id: "Dumbbell Overhead Press", category: .press),
        // Pull
        Entry(id: "Barbell Row",           category: .pull),
        Entry(id: "Pendlay Row",           category: .pull),
        Entry(id: "Pull-Up",               category: .pull),
        Entry(id: "Weighted Pull-Up",      category: .pull),
        Entry(id: "Lat Pulldown",          category: .pull),
        Entry(id: "Cable Row",             category: .pull),
        // Carry
        Entry(id: "Farmers Carry",         category: .carry),
        Entry(id: "Suitcase Carry",        category: .carry),
        // Olympic Weightlifting
        Entry(id: "Squat Snatch",          category: .olympic),
        Entry(id: "Squat Clean",           category: .olympic),
        Entry(id: "Power Clean",           category: .olympic),
        Entry(id: "Power Snatch",          category: .olympic),
        Entry(id: "Hang Clean",            category: .olympic),
        Entry(id: "Hang Snatch",           category: .olympic),
        Entry(id: "Split Jerk",            category: .olympic),
        Entry(id: "Push Jerk",             category: .olympic),
        Entry(id: "Clean and Jerk",        category: .olympic),
    ]

    public static let exerciseIDs: Set<String> = Set(all.map(\.id))
    public static let defaultExerciseID: String = all.first!.id

    public static var sortedByCategory: [Entry] {
        let order: [Category] = [.squat, .hinge, .press, .pull, .carry, .olympic]
        return order.flatMap { cat in
            all.filter { $0.category == cat }.sorted { $0.id < $1.id }
        }
    }

    public static func isWeightliftingExercise(_ exerciseID: String) -> Bool {
        exerciseIDs.contains(exerciseID)
    }
}

public enum ConditioningScoringType: String, Codable, CaseIterable, Sendable {
    case time = "time"
    case reps = "reps"

    public func isBetterThan(newValue: Double, existingValue: Double?) -> Bool {
        guard let existing = existingValue else { return true }
        switch self {
        case .time: return newValue < existing
        case .reps: return newValue > existing
        }
    }

    public func formatValue(_ value: Double) -> String {
        switch self {
        case .time:
            let totalSeconds = Int(value)
            let minutes = totalSeconds / 60
            let seconds = totalSeconds % 60
            return "\(minutes):\(String(format: "%02d", seconds))"
        case .reps:
            let count = Int(value)
            return count == 1 ? "1 rep" : "\(count) reps"
        }
    }
}

public enum ConditioningExerciseCatalog {

    public struct Entry: Identifiable, Sendable {
        public let id: String
        public let defaultScoringType: ConditioningScoringType
        public init(id: String, defaultScoringType: ConditioningScoringType) {
            self.id = id; self.defaultScoringType = defaultScoringType
        }
    }

    public static let all: [Entry] = [
        // Reps-based
        Entry(id: "Wall Ball",               defaultScoringType: .reps),
        Entry(id: "Box Jump",                defaultScoringType: .reps),
        Entry(id: "Burpee",                  defaultScoringType: .reps),
        Entry(id: "Kettlebell Swing",        defaultScoringType: .reps),
        Entry(id: "Double Under",            defaultScoringType: .reps),
        Entry(id: "Pull-Up (Kipping)",       defaultScoringType: .reps),
        Entry(id: "Toes-to-Bar",             defaultScoringType: .reps),
        Entry(id: "Muscle-Up",               defaultScoringType: .reps),
        Entry(id: "Push-Up",                 defaultScoringType: .reps),
        Entry(id: "Sit-Up",                  defaultScoringType: .reps),
        Entry(id: "Air Squat",               defaultScoringType: .reps),
        Entry(id: "Thruster",                defaultScoringType: .reps),
        Entry(id: "Rowing (Calories)",       defaultScoringType: .reps),
        Entry(id: "Assault Bike (Calories)", defaultScoringType: .reps),
        // Time-based
        Entry(id: "400m Run",                defaultScoringType: .time),
        Entry(id: "800m Run",                defaultScoringType: .time),
        Entry(id: "1-Mile Run",              defaultScoringType: .time),
        Entry(id: "5K Run",                  defaultScoringType: .time),
        Entry(id: "500m Row",                defaultScoringType: .time),
        Entry(id: "2K Row",                  defaultScoringType: .time),
        Entry(id: "1K Assault Bike",         defaultScoringType: .time),
    ]

    public static let exerciseIDs: Set<String> = Set(all.map(\.id))

    public static func isConditioningExercise(_ exerciseID: String) -> Bool {
        exerciseIDs.contains(exerciseID)
    }

    public static func scoringType(for exerciseID: String) -> ConditioningScoringType? {
        all.first { $0.id == exerciseID }?.defaultScoringType
    }
}

File: wod-dashboard/src/lib/exercise-catalog.ts
export interface WeightliftingExercise {
  name: string;
  category:
    | "Squat"
    | "Hip Hinge"
    | "Press"
    | "Pull"
    | "Carry"
    | "Olympic Weightlifting";
}

export interface ConditioningExercise {
  name: string;
  scoring: "Reps" | "Time";
}

export const weightliftingExercises: WeightliftingExercise[] = [
  // Squat (6)
  { name: "Back Squat", category: "Squat" },
  { name: "Front Squat", category: "Squat" },
  { name: "Safety Bar Squat", category: "Squat" },
  { name: "Box Squat", category: "Squat" },
  { name: "Pause Squat", category: "Squat" },
  { name: "Goblet Squat", category: "Squat" },

  // Hip Hinge (10)
  { name: "Conventional Deadlift (No Straps)", category: "Hip Hinge" },
  { name: "Conventional Deadlift (With Straps)", category: "Hip Hinge" },
  { name: "Romanian Deadlift (No Straps)", category: "Hip Hinge" },
  { name: "Romanian Deadlift (With Straps)", category: "Hip Hinge" },
  { name: "Sumo Deadlift (No Straps)", category: "Hip Hinge" },
  { name: "Sumo Deadlift (With Straps)", category: "Hip Hinge" },
  { name: "Trap Bar Deadlift (No Straps)", category: "Hip Hinge" },
  { name: "Trap Bar Deadlift (With Straps)", category: "Hip Hinge" },
  { name: "Good Morning", category: "Hip Hinge" },
  { name: "Hip Thrust", category: "Hip Hinge" },

  // Press (6)
  { name: "Flat Barbell Bench Press", category: "Press" },
  { name: "Incline Barbell Bench Press", category: "Press" },
  { name: "Strict Press", category: "Press" },
  { name: "Push Press", category: "Press" },
  { name: "Dumbbell Bench Press", category: "Press" },
  { name: "Dumbbell Overhead Press", category: "Press" },

  // Pull (6)
  { name: "Barbell Row", category: "Pull" },
  { name: "Pendlay Row", category: "Pull" },
  { name: "Pull-Up", category: "Pull" },
  { name: "Weighted Pull-Up", category: "Pull" },
  { name: "Lat Pulldown", category: "Pull" },
  { name: "Cable Row", category: "Pull" },

  // Carry (2)
  { name: "Farmers Carry", category: "Carry" },
  { name: "Suitcase Carry", category: "Carry" },

  // Olympic Weightlifting (9)
  { name: "Squat Snatch", category: "Olympic Weightlifting" },
  { name: "Squat Clean", category: "Olympic Weightlifting" },
  { name: "Power Clean", category: "Olympic Weightlifting" },
  { name: "Power Snatch", category: "Olympic Weightlifting" },
  { name: "Hang Clean", category: "Olympic Weightlifting" },
  { name: "Hang Snatch", category: "Olympic Weightlifting" },
  { name: "Split Jerk", category: "Olympic Weightlifting" },
  { name: "Push Jerk", category: "Olympic Weightlifting" },
  { name: "Clean and Jerk", category: "Olympic Weightlifting" },
];

export const conditioningExercises: ConditioningExercise[] = [
  // Reps (14)
  { name: "Wall Ball", scoring: "Reps" },
  { name: "Box Jump", scoring: "Reps" },
  { name: "Burpee", scoring: "Reps" },
  { name: "Kettlebell Swing", scoring: "Reps" },
  { name: "Double Under", scoring: "Reps" },
  { name: "Pull-Up (Kipping)", scoring: "Reps" },
  { name: "Toes-to-Bar", scoring: "Reps" },
  { name: "Muscle-Up", scoring: "Reps" },
  { name: "Push-Up", scoring: "Reps" },
  { name: "Sit-Up", scoring: "Reps" },
  { name: "Air Squat", scoring: "Reps" },
  { name: "Thruster", scoring: "Reps" },
  { name: "Rowing (Calories)", scoring: "Reps" },
  { name: "Assault Bike (Calories)", scoring: "Reps" },

  // Time (7)
  { name: "400m Run", scoring: "Time" },
  { name: "800m Run", scoring: "Time" },
  { name: "1-Mile Run", scoring: "Time" },
  { name: "5K Run", scoring: "Time" },
  { name: "500m Row", scoring: "Time" },
  { name: "2K Row", scoring: "Time" },
  { name: "1K Assault Bike", scoring: "Time" },
];

export const allExerciseNames: string[] = [
  ...weightliftingExercises.map((e) => e.name),
  ...conditioningExercises.map((e) => e.name),
];
